docs/tutorials/security/part-2.md

```diff
@@ -0,0 +1,266 @@
+- Strings: as there is no limit on their lengths
+- Lists, sets, maps: that can contain an arbitrary number of items
+
+&rarr; **SOLUTION**:
+
+- Ask the user to pay a minimum tez for each call
+- Set a threshold limit
+- Store data on a big_map
+- Avoid unnecessary onchain computation that can be done offchain. Ex: do not loop onchain and just update a part of a map
+
+Example with the current FA1.2 implementation: https://inference.ag/blog/2023-10-09-FA12_spenders/
+
+You can have a look at the LIGO implementation of fa1.2 on the Ligo registry [here](https://packages.ligolang.org/package/ligo_fa1.2)
+
+The code follows the standard but you can see that the [Allowance type is a map](https://github.com/frankhillard/ligoFA12/blob/main/lib/asset/allowance.mligo#L3C8-L3C8). It would have been better to change the Standard and use a `big_map` instead of a `map`. If you implement the Standard differently, then your smart contract storage definition and entrypoint signatures will not match anymore and will not be supported by other platforms
+
+## Re-entrancy
+
+These attacks allow an attacker to repeatedly call a contract function in a way that drains the contract’s resources, leading to a denial of service (DoS) attack
+
+One of the most well-known examples of a re-entrancy attack occurred in 2016 when an attacker exploited a vulnerability in the DAO (Decentralized Autonomous Organization) contract on the Ethereum blockchain. But this popular hack is still actively used :
+
+- Uniswap/Lendf.Me hacks (April 2020) – $25 mln, attacked by a hacker using a re-entrancy.
+- The BurgerSwap hack (May 2021) – $7.2 mln, because of a fake token contract and a re-entrancy exploit.
+- The SURGEBNB hack (August 2021) – $4 mln, seems to be a re-entrancy-based price manipulation attack.
+- CREAM FINANCE hack (August 2021) – $18.8 mln, re-entrancy vulnerability allowed the exploiter for the second borrow.
+- Siren protocol hack (September 2021) – $3.5 mln, AMM pools were exploited through re-entrancy attack.
+
+This kind of attack is quite simple to put in place with Solidity because of the way it works.
+
+Consider this scenario :
+
+```mermaid
+sequenceDiagram
+  User->>MaliciousContract: deposit funds
+  MaliciousContract->>LedgerContract: deposit funds
+  User->>MaliciousContract: call withdraw
+  MaliciousContract->>LedgerContract: call withdraw
+  Note right of LedgerContract: checkBalance
+  Note right of LedgerContract: sendFunds
+  LedgerContract->>MaliciousContract: sendFunds operation
+  Note right of MaliciousContract: loop calling withdraw ... x times
+  MaliciousContract->>LedgerContract: call withdraw
+  LedgerContract->>MaliciousContract: sendFunds operation  ... x times
+  Note right of LedgerContract: ... Once finish ... UpdateBalance
+```
+
+Why this scenario is not possible on Solidity?
+On Solidity, the operation will call directly the smart contract like doing a stop, calling a synchronous execution and continuing the flow.
```

Comment:
Not sure if I follow these two sentences. Do you mean "Why is this scenario possible on Solidity"? 

I can't follow "doing a stop, calling a synchronous execution, and..." Can explain more clearly what the flow is here? Are you saying that EVM batches transactions which can trick it into running transactions multiple times in error?

Reply:
yes, typo on the **not**